### PR TITLE
Don't offer to inline variables if it would cause them to no longer be definitely assigned.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/15408

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

File: src/Features/CSharp/Portable/InlineDeclaration/CSharpInlineDeclarationDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -200,6 +214,100 @@ private void AnalyzeSyntaxNode(SyntaxNodeAnalysisContext context)
...
+                                                            .Select(n => n.Span)
+                                                            .Distinct();
+
+            return diagnosticSpans.Intersect(newReferenceSpans).Any();
+
+            // var annotation = new SyntaxAnnotation();
+            // var declaratorWithoutInitializer = rootWithoutInitializer.GetAnnotatedNodes(annotation)
+            //                                                         .Single();
+            // var currentLocalSymbol = (ILocalSymbol)newSemanticModel.GetDeclaredSymbol(declaratorWithoutInitializer);
+#if false
```
Comment (jaredpar) [inline src/Features/CSharp/Portable/InlineDeclaration/CSharpInlineDeclarationDiagnosticAnalyzer.cs]: Dead code?

File: src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs [hunk 2]
```diff
@@ -21,6 +21,8 @@ internal class CSharpSyntaxGenerator : SyntaxGenerator
     {
         internal override SyntaxTrivia CarriageReturnLineFeed => SyntaxFactory.CarriageReturnLineFeed;
 
+        public static readonly SyntaxGenerator Instance = new CSharpSyntaxGenerator();
```
Comment (jaredpar) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: We should unify this with the similar pattern in CSharpFlagsEnumGenerator

- http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Workspaces/CodeGeneration/CSharpFlagsEnumGenerator.cs,23
- http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Workspaces/CodeGeneration/CSharpFlagsEnumGenerator.cs,44

Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs]: Sure.

File: src/Features/CSharp/Portable/InlineDeclaration/CSharpInlineDeclarationDiagnosticAnalyzer.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/InlineDeclaration/CSharpInlineDeclarationDiagnosticAnalyzer.cs]: Code that i would like to write, but cannot until https://github.com/dotnet/roslyn/issues/15430 completes.

File: src/Workspaces/Core/Portable/Editing/SyntaxEditor.cs [hunk 3]
```diff
@@ -29,6 +29,13 @@ public SyntaxEditor(SyntaxNode root, Workspace workspace)
             _changes = new List<Change>();
         }
 
+        internal SyntaxEditor(SyntaxNode root, SyntaxGenerator generator)
```
Comment (dpoeschl) [inline src/Workspaces/Core/Portable/Editing/SyntaxEditor.cs]: It doesn't look like this is used anywhere. Am I missing something?

File: src/Features/CSharp/Portable/InlineDeclaration/CSharpInlineDeclarationDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -206,6 +217,79 @@ private void AnalyzeSyntaxNode(SyntaxNodeAnalysisContext context)
...
+
+            var references = query.ToImmutableArray<SyntaxNode>();
+
+            var root = semanticModel.SyntaxTree.GetCompilationUnitRoot(cancellationToken);
+
+            // Ensure we can track the references and the local variable as we make edits
+            // to the tree.
+            var rootWithTrackedNodes = root.TrackNodes(references.Concat(localDeclarator).Concat(enclosingBlock));
+
+            // Now, take the local variable and remove it's initializer.  Then go to all
```
Comment (dpoeschl) [inline src/Features/CSharp/Portable/InlineDeclaration/CSharpInlineDeclarationDiagnosticAnalyzer.cs]: it's ==> its

**Pull request merged**